### PR TITLE
Write QuickCheck docs

### PR DESCRIPTION
**Pull request by annenkov**
## Purpose

Add QuickCheck documentation

### CODE REVIEW

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 1]
```diff
@@ -482,11 +482,169 @@ account:
         let amount1 = Amount::from_ccd(20);
         claim_eq!(host.get_transfers_to(receiver0), [amount0, amount1]);
 
+.. _writing_property_based_tests:
+
+Writing property-based tests
+============================
+
+Property-based testing technique allows for testing statements about your code that expected to be true for any input parameters.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:490]:
```suggestion
The property-based testing technique allows for testing statements about your code that are expected to be true for any input parameters.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 2]
```diff
@@ -482,11 +482,169 @@ account:
...
+
+By default, this command compiles the contract, unit tests and QuickCheck tests to machine code for your local target (most likely x86_64), and runs them.
+
+.. note::
+
+    Printing and supplying a seed is only possible using ``cargo concordium test``.
+
+.. warning::
+
+    Avoid using ``clam!`` and ``claim_eq!`` in ``#[concordium_quickcheck]`` tests.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:559]: Perhaps something like:

```suggestion
    Avoid using ``claim!``, ``assert_eq!``, ``panic!`` or any other command that panics in ``#[concordium_quickcheck]`` tests.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 3]
```diff
@@ -482,11 +482,169 @@ account:
...
 
+.. _writing_property_based_tests:
+
+Writing property-based tests
+============================
+
+Property-based testing technique allows for testing statements about your code that expected to be true for any input parameters.
+The input to such tests is generated randomly.
+Property-based testing is supported using the |QuickCheck|_ crate.
+The tests should placed in the same module as Wasm unit tests and annotated with the ``#[concordium_quickcheck]`` macro.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:493]:
```suggestion
The tests should be placed in the same module as regular unit tests and annotated with the ``#[concordium_quickcheck]`` macro.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 4]
```diff
@@ -482,11 +482,169 @@ account:
...
+The macro takes a named attribute ``num_tests`` for specifying the number of random tests to run.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:494]: Consider moving this line to the example, as that gives more context for the user to understand it. The user might not know what a named attribute is in a macro. But seeing it with the example makes it quite obvious.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 5]
```diff
@@ -482,11 +482,169 @@ account:
...
+If no ``num_tests`` given, the default number is ``100``.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:495]:
```suggestion
If no ``num_tests`` is given, the default number is ``100``.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 6]
```diff
@@ -482,11 +482,169 @@ account:
...
+The return value of the function should be a boolean value with ``true`` corresponding to the fact that the property holds.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:496]:
```suggestion
The return value of the function should be a boolean corresponding to whether the property holds.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 7]
```diff
@@ -482,11 +482,169 @@ account:
...
+In ``Cargo.toml``, ``concordium-std`` should be specified as a ``dev`` dependency with the ``concordium-quickcheck`` feature enabled.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:497]:
```suggestion
 (newline)
To get started, add the `concordium-quickcheck` feature to `concordium-std` as a `dev-dependency` In ``Cargo.toml``:
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 8]
```diff
@@ -482,11 +482,169 @@ account:
...
+       #[concordium_quickcheck(num_tests = 500)]
+       fn some_property_test(address: Address, amount: Amount) -> bool {
+        ...
+        // Instantiate custom struct with random parameters.
+        let input = MyParameters { sender: address, payment: amount }
+        ...
+        }
+    }
+
+The type like ``Address`` and ``Amount`` in the example have ``Arbitrary`` trait implementations, which are used to obtain random values.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:525]:
```suggestion
The types ``Address`` and ``Amount`` in the example have ``Arbitrary`` trait implementations, which are used to obtain random values.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 9]
```diff
@@ -482,11 +482,169 @@ account:
...
+.. code-block:: rust
+
+    #[concordium_cfg_test]
+    mod test {
+
+       #[concordium_quickcheck(num_tests = 500)]
+       fn some_property_test(address: Address, amount: Amount) -> bool {
+        ...
+        // Instantiate custom struct with random parameters.
+        let input = MyParameters { sender: address, payment: amount }
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:520]: It is a bit unclear that this isn't always needed.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 10]
```diff
@@ -482,11 +482,169 @@ account:
...
+        ...
+        }
+    }
+
+The type like ``Address`` and ``Amount`` in the example have ``Arbitrary`` trait implementations, which are used to obtain random values.
+Read more about available ``Arbitrary`` instances for Concordium-specific type in |concordium_contracts_common|_ documentation.
+``Arbitrary`` instances for standard data types, like numbers and collections (``Vec``, ``BTreeMap``) are defined in |QuickCheck|_ and available automatically.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:527]: Perhaps?
```suggestion
``Arbitrary`` instances for standard data types, like numbers and collections (``Vec``, ``BTreeMap``) are defined in |QuickCheck|_ and available by default.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 11]
```diff
@@ -482,11 +482,169 @@ account:
...
+Custom user data type instances can be created directly in tests using the random input parameters or by defining ``Arbitrary`` instances.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:528]: You should mention the `MyParameters` here.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 12]
```diff
@@ -482,11 +482,169 @@ account:
...
+
+The same command is used for running Wasm QuickCheck tests as in :ref:`tests_in_wasm`:
+
+.. code-block:: console
+
+    cargo concordium test
+
+When a test fails, it reports the random seed used to produce the input values.
+After making the required fixes one can use the same seed to see whether the tests works on the same generated values.
+The seed is a ``u64`` number, which can be provided along with the test command.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:539]:
```suggestion
The seed is a ``u64`` number, which can be provided along with the test command:
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 13]
```diff
@@ -482,11 +482,169 @@ account:
...
+
+.. code-block:: console
+
+    cargo concordium test --seed 1234567890
+
+Concordium QuickCheck tests can also be run with
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:545]:
```suggestion
Concordium QuickCheck tests can also be run with:
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 14]
```diff
@@ -482,11 +482,169 @@ account:
...
+    Printing and supplying a seed is only possible using ``cargo concordium test``.
+
+.. warning::
+
+    Avoid using ``clam!`` and ``claim_eq!`` in ``#[concordium_quickcheck]`` tests.
+    Using these makes tests fail without providing a counterexample when running tests with ``cargo concordium test``.
+    Return a boolean value instead.
+
+
+Consider the following example.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:564]: Consider adding a sub-section, bold text, or similar for this example. Or at least make it clear that it is different from the general info about quickcheck.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 15]
```diff
@@ -482,11 +482,169 @@ account:
...
+
+If we change the highlighted lines in the code above to
+
+.. code-block:: rust
+
+    if self.count <= self.threshold {
+                self.count += 1;
+            }
+
+The test fails with a counterexample (that is, an input that breaks the property).
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:628]:
```suggestion
The test fails with a counterexample, i.e., an input that breaks the property:
```

**Review by Bargsteen — COMMENTED** [review]
Nicely explained. I have a number of smaller points, some reordering and grammar, but overall it is good!

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 9 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:520]: Ok, I'll add "if necessary" :)

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 16]
```diff
@@ -482,11 +482,169 @@ account:
...
+        fn new(threshold: u32) -> Self {
+            State {
+                count: 0,
+                threshold,
+            }
+        }
+
+        // Increment only if the current count is below the threshold.
+        fn increment(&mut self) {
+            if self.count < self.threshold {
```
[the remark this answers is not in the document]
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:590]: Yes, I think it makes sense to reverse the presentation.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 2 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:561]: I think `clam!` and `claim_eq!` are "special": they don't just panic when running with `cargo-concordium`, they call `ReportError`. Using `assert` in Wasm would just give `Unreachable`.

I'll clarify this here as well. Because `claim!` will produce a user defined error message and interrupt the test. So, the counterexample won't be printed.

Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst]: Fixed now.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 14 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst]: Done.

**Comment by annenkov** [comment]
@Bargsteen thanks for your comments! I've addressed your review.

**Comment by annenkov** [comment]
@Bargsteen thanks for your comments! I've addressed your review.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 17]
```diff
@@ -482,11 +482,189 @@ account:
...
+.. code-block::
+
+    ...
+
+    [dev-dependencies]
+    concordium-std = { version = "5", features = ["concordium-quickcheck"] }
+
+    ...
+
+The ``concordium_quickcheck`` macro takes a named attribute ``num_tests`` for specifying the number of random tests to run.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:507]:
```suggestion
The ``concordium_quickcheck`` macro takes the ``num_tests`` attribute for specifying the number of random tests to run.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 18]
```diff
@@ -482,11 +482,189 @@ account:
...
+In the code snipped below, the parameters ``address`` and ``amount`` are generated randomly.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:508]:
```suggestion
In the code snippet below, the parameters ``address`` and ``amount`` are generated randomly.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 19]
```diff
@@ -482,11 +482,189 @@ account:
...
+See more details on QuickCheck's ``Arbitrary`` `here <https://docs.rs/quickcheck/latest/quickcheck/trait.Arbitrary.html>`_.
+
+The same command is used for running Wasm QuickCheck tests as in :ref:`tests_in_wasm`:
+
+.. code-block:: console
+
+    $cargo concordium test
+
+When a test fails, it reports the random seed used to produce the input values.
+After making the required fixes to the code, one can use the same seed to see whether the previously failed tests work on the same generated values.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:539]:
```suggestion
After making the required fixes to the code, you can use the same seed to see whether the previously failed tests work on the same generated values.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 20]
```diff
@@ -482,11 +482,189 @@ account:
...
+
+.. note::
+
+    Printing and supplying a seed is only possible using ``cargo concordium test``.
+
+.. warning::
+
+    Avoid using ``fail!`` and ``claim!`` variants in ``#[concordium_quickcheck]`` tests.
+    In Wasm unit tests (see :ref:`tests_in_wasm`) these commands report an error.
+    However, using them in QuickCheck tests, makes the tests fail without providing a counterexample when running with ``cargo concordium test``.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:562]:
```suggestion
    However, using them in QuickCheck tests makes the tests fail without providing a counterexample when running with ``cargo concordium test``.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 21]
```diff
@@ -482,11 +482,189 @@ account:
...
+    Avoid also using ``assert_eq!``, ``panic!`` or any other command that panics.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:563]:
```suggestion
    Also avoid using ``assert_eq!``, ``panic!`` or any other command that panics.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 22]
```diff
@@ -482,11 +482,189 @@ account:
...
+    Return a boolean value instead.
+
+Example
+-------
+
+Consider a counter with a threshold: if the count is less then the threshold, it gets incremented, or stays unchanged otherwise.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:569]:
```suggestion
Consider a counter with a threshold: if the count is less than the threshold, it gets incremented; otherwise, it stays unchanged.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 23]
```diff
@@ -482,11 +482,189 @@ account:
...
+.. code-block::
+
+  TestResult {
+    status: Fail,
+    arguments: [
+        "0",
+    ],
+    err: None,
+
+The ``arguments`` part shows the values that caused the test to fails.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:636]:
```suggestion
The ``arguments`` part shows the values that caused the test to fail.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 24]
```diff
@@ -482,11 +482,189 @@ account:
...
+In this case, if the threshold is ``0``, then the counter becomes ``1`` after calling ``state.increment()`` breaking the property.
+
+.. note::
+
+    |QuickCheck|_ implements a special mechanism called "shrinking" to find a simplest counterexample.
+    For the example above, ``0`` is the simplest input on which the test failed.
+
+If we change the highlighted lines in the code above to
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:644]:
```suggestion
If you change the highlighted lines in the code above to
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 25]
```diff
@@ -482,11 +482,189 @@ account:
...
+        self.count += 1;
+    }
+
+Then all ``500`` tests pass successfully.
+
+.. note::
+
+    The fact that many random tests passed successfully does not automatically mean that the property holds for **all** inputs.
+    Often the input space is quite large to be covered fully.
+    In this case, it is importnat to think carefully what an implementation the ``Arbitrary`` trait is doing to generate random input for your specific data.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:658]:
```suggestion
    In this case, it is important to think carefully about what an implementation of the ``Arbitrary`` trait is doing to generate random input for your specific data.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 26]
```diff
@@ -482,11 +482,189 @@ account:
...
+    In order to cover corner cases, one can bias the generated data to produce values that a deemed as potentially problematic.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:659]:
```suggestion
    In order to cover corner cases, you can bias the generated data to produce values that are deemed as potentially problematic.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 27]
```diff
@@ -482,11 +482,189 @@ account:
...
+
+  TestResult {
+    status: Fail,
+    arguments: [
+        "0",
+    ],
+    err: None,
+
+The ``arguments`` part shows the values that caused the test to fails.
+In this case, if the threshold is ``0``, then the counter becomes ``1`` after calling ``state.increment()`` breaking the property.
```
Comment (dg-concordium) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:637]:
```suggestion
In this case, if the threshold is ``0``, then the counter becomes ``1`` after calling ``state.increment()``, breaking the property.
```

**Review by dg-concordium — APPROVED** [review]

**Comment by annenkov** [comment]
@dg-concordium thanks for your comments!

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 28]
```diff
@@ -482,11 +482,193 @@ account:
         let amount1 = Amount::from_ccd(20);
         claim_eq!(host.get_transfers_to(receiver0), [amount0, amount1]);
 
+.. _writing_property_based_tests:
+
+Writing property-based tests
+============================
+
+The property-based testing technique allows for testing statements about your code that are expected to be true for any input parameters.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:490]: I am not sure this is very meaningful. Quite often one assumes some relationship between the inputs (i.e., ensure some precondition). So this phrasing is a bit misleading in my view.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 29]
```diff
@@ -482,11 +482,193 @@ account:
...
+The input to such tests is generated randomly.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:491]: Could you emphasize here what randomly entails. That one has to write a generator that generates meaningful inputs. And that the efficacy of the testing is heavily dependent on this.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 30]
```diff
@@ -482,11 +482,193 @@ account:
...
+Custom user data type instances, like ``MyParameters`` above, can be created directly in tests using the random input parameters or by defining ``Arbitrary`` instances.
+See more details on QuickCheck's ``Arbitrary`` `here <https://docs.rs/quickcheck/latest/quickcheck/trait.Arbitrary.html>`_.
+
+The same command is used for running Wasm QuickCheck tests as in :ref:`tests_in_wasm`:
+
+.. code-block:: console
+
+    $cargo concordium test
+
+When a test fails, it reports the random seed used to produce the input values.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:538]: Please explain here that all the random inputs are generated deterministically from the random seed. That is implied here, but it would be good to make it explicit for non-experts.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 31]
```diff
@@ -482,11 +482,193 @@ account:
...
+        ],
+        err: None,
+    }
+
+The ``arguments`` part shows the values that caused the test to fail.
+In this case, if the threshold is ``0`` and the number of calls is ``1``, then the counter becomes ``1`` after calling ``state.increment()``, breaking the property.
+
+.. note::
+
+    |QuickCheck|_ implements a special mechanism called "shrinking" to find a simplest counterexample.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:645]:
```suggestion
    |QuickCheck|_ implements a special mechanism called "shrinking" to find the simplest counterexample.
```

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 32]
```diff
@@ -482,11 +482,193 @@ account:
...
+    Avoid using ``fail!`` and ``claim!`` variants in ``#[concordium_quickcheck]`` tests.
+    In Wasm unit tests (see :ref:`tests_in_wasm`) these commands report an error.
+    However, using them in QuickCheck tests makes the tests fail without providing a counterexample when running with ``cargo concordium test``.
+    Also avoid using ``assert_eq!``, ``panic!`` or any other command that panics.
+    Return a boolean value instead.
+
+Example
+-------
+
+Consider a counter with a threshold: if the count is less than the threshold, it gets incremented; otherwise, it stays unchanged.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:569]: While this example is good warm-up example, I would prefer if we have a more realistic one.

Is there something in the piggy-bank example that we could specify like this?

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 28 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:490]: If one has a precondition then the property is an implication. When the precondition is not met, the test is discarded (that is, holds vacuously). 
I can make it more explicit that for any input that possibly satisfies certain validity criteria. Would it be better?

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 29 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:491]: I have a note at the end of the docs after the example. Should I move it here?

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 32 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:569]: I looked at `piggy-bank` with exactly this purpose. But all I could come up with looks as artificial as this example. However, once we integration testing, we could make a nice story about `piggy-bank` being a transition system and show some temporal properties, e.g. reachability of states over an arbitrary sequence of calls.

wCCD could be a possibility, but these examples would be more complex and require more background. I wanted to make a separate task of adding property tests to existing examples and then writing up some sort of tutorial, maybe explaining the methodology in more detail and showing how to write custom `Arbitrary`. Does it make sense?

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 30 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:538]: Done.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 28 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst]: I've added an explanation, a general "template" and an example of a property.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 32 above)
Comment (Bargsteen) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:574]: While the current example is contrived I think it shows enough info to get started **if** you know something about property-based testing already. Perhaps we should link to some general information about property-based testing. Do you know any good beginner resources?

And I think a separate tutorial is a good idea.

Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:574]: I don't think I have enough time to pick (or evaluate) good resources on property-based testing, if we plan to finish up with the PR today-ish. I can add *some* links, of course, just as initial pointers.

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 29 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:493]: I converted a note into a warning a moved it right after I write about `Arbitrary`

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst (hunk 32 above)
Comment (Bargsteen) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:583]: Ok, the link was just a suggestion, we can do it later 😊

Comment (Bargsteen) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:583]: Is it ready for another review?

Comment (annenkov) [reply source/mainnet/smart-contracts/guides/unit-test-contract.rst:583]: @Bargsteen yep, requested re-review :)

**Review by abizjak — APPROVED** [review]

File: source/mainnet/smart-contracts/guides/unit-test-contract.rst [hunk 33]
```diff
@@ -482,11 +482,200 @@ account:
...
+.. _writing_property_based_tests:
+
+Writing property-based tests
+============================
+
+The property-based testing technique allows for testing statements about your code that are expected to be true for any input parameters, possibly satisfying some precondition.
+You can think of a precondition and a property as functions returning a boolean.
+That is, for a function ``fun``, a property looks as the following: "for any input ``x``, ``y``, ``z``, such that ``precondition(x, y, z) = true``, ``property(x, y, z, fun(x,y,z)) = true``".
+The input to such tests is generated randomly.
+An example of a property is "for any integers ``n`` and ``m``, such that ``even(n) = true`` and ``even(m) = true``, ``even(n + m) = true``".
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/guides/unit-test-contract.rst:494]: This section is a great addition! Very concise and easy to understand. Nice work ;)
